store/dhash/dhash.go

```diff
@@ -64,7 +67,7 @@ func EncryptAES(payload, passphrase []byte) ([]byte, []byte, error) {
 	// Create initialization vector (nonse) to be used during encryption
 	// Nonce is derived from the mulithash (passpharase) so that encrypted payloads
 	// for the same multihash can be compared to each other without having to decrypt
-	nonce := SHA256(passphrase, nil)[:nonceLen]
+	nonce := SHA256(append(noncePrefix, passphrase...), nil)[:nonceLen]
```

Comment:
AFAIT this means that two different peer will use the same nonce (since it is just a salted hash of the multihash). Assuming this is true:

This is not OK, because then an attacker can use multiple different provider records from multiple different nodes using the same nonce to recover the derived key, which is the multihash. (the actual attack relies on cipher textes being different, here the cipher text should be the peer id)
If this what the spec says, the spec needs to be updated.

You can search: `aes gcm duplicated nonce` this is a well known attack.
Other symetric encryptions constructions supports using duplicated nonce with minimal issue (attackers can match duplicates without revealing their content, either based on blocks of the stream without feedback, or until the first duplicate is found if there is feedback).
We are not using this here tho, and this is unsafe.

cc @guillaumemichel we had brainstorming together around this, I was maybe not clear at the end.

Reply:
Responding to:
> We derive nonse from the passphrase so that we can deterministically produce the same result of `hash(mh) -> EncPeerID` given the same multihash / peerID.

You need to salt this by the peer id then:
```go
nonce := SHA256(append(append(binary.AppendUint64(noncePrefix, uint64(len(payload))), payload...)), passphrase...))
```

Reply:
Thanks for pointing this out - will push a fix in shortly